### ISSUE TITLE
question about data and bc

### ISSUE THREAD

**Issue by xy1015874723** [lululxvi/deepxde#257]
hi, lu 
DeepXDE is really nice, and i have 2 questions.
when i use DeepXDE, can i use my own data as training data. if i can ,how to import it?
Can i use DeepXDE describe bc like this u(0,t)=finite?
i am looking forward your response, thank you very much

**Comment by lululxvi** [comment]
1. Yes, you can. Use the `anchors` argument when defining PDE, see https://deepxde.readthedocs.io/en/latest/modules/deepxde.data.html#deepxde.data.pde.PDE
2. What do you mean by "finite"?

**Comment by xy1015874723** [comment]
Thank you, lu.
finite like this PDE bc,ic
![image](https://user-images.githubusercontent.com/73056004/114291596-46a96300-9abb-11eb-8c6b-ad40972cdff8.png)
which can be used to determine a coefficient of solution

**Comment by xy1015874723** [comment]
I am sorry to bother you again. I am still a little confused about that if  i can use DeepXDE  training my data which solution is also given by me, not just given space-time locations.

**Comment by xy1015874723** [comment]
@lululxvi  I met this error when i use anchors. Could you please give me some suggestions , and i will be appreciate it. Here are the error and my code 
![image](https://user-images.githubusercontent.com/73056004/114526706-cf6dfd80-9c79-11eb-886f-f191aa4f90d1.png)
![image](https://user-images.githubusercontent.com/73056004/114526736-d5fc7500-9c79-11eb-8819-1f18fd1a3fff.png)

**Comment by lululxvi** [comment]
> I am sorry to bother you again. I am still a little confused about that if  i can use DeepXDE  training my data which solution is also given by me, not just given space-time locations. 

Check the examples of inverse problems.

**Comment by xy1015874723** [comment]
@lululxvi ok, i will check the examples, thanks.
I have solved the error by setting num_boundary nonzero , and it works. But i don't know the reason of the error.

**Comment by xy1015874723** [comment]
I am sorry to bother you. I have learned inverse examples, and here is my code. 
![image](https://user-images.githubusercontent.com/73056004/115247311-42371700-a159-11eb-87c7-c8d324a87aff.png)
Luckily, the results seems correct in most of the domain. However, it is far from the exact solution somewhere, and i am sure my the solution(value) given by me is exact.
so, do my code also have mistakes and need to improve?  if the code is correct, i will look for reasons in other ways. Could you give me some advice? Thanks a lot.

**Comment by xy1015874723** [comment]
shape of arrays
ob_x, array (100,1)
ob_t, array(11,1)
ob_y, array(100, 11)
ob_yy, array(1100,1)

**Comment by lululxvi** [comment]
Your code looks good. It is wired. I need more information to find what happened here.

**Comment by xy1015874723** [comment]
@lululxvi I thank this information may be useful.
(a)
![image](https://user-images.githubusercontent.com/73056004/115953624-3368af80-a51f-11eb-940f-eb0f79d6b504.png)
![image](https://user-images.githubusercontent.com/73056004/115954563-55b0fc00-a524-11eb-99f5-a49599166d7f.png)
(b)
![image](https://user-images.githubusercontent.com/73056004/115954249-8001ba00-a522-11eb-8df0-ceeca1834794.png)
(c)
![image](https://user-images.githubusercontent.com/73056004/115954253-8859f500-a522-11eb-82d2-768fc196d87e.png)
(a) is the exact solution where I generate training data. As mentioned above,
ob_x, array (100,1) (0.01,1, 0.01)
ob_t, array(11,1)  (0, 0.5, 0.05)
ob_y, array(100, 11)
ob_yy, array(1100,1)
(b) and (c) is the training result,  and the results seems correct in most of the domain. But in t=0.05, it is obvious  Concave which is confusing and incorrect(refer to t=0.0574 in a), but I don't know the reason.  Code is same as what  I write above.
If you know the reason, please tell me.
Thanks.

**Comment by lululxvi** [comment]
Is your right BC correct? In your figure, it seems a Dirichelet BC, but in the code it is Neumann.

**Comment by xy1015874723** [comment]
@lululxvi Oh ,really sorry. Actually， I trained two models, and I met the same problem. I took a mistake when uploaded pictures.
Here are the results which are given by the code above.
![image](https://user-images.githubusercontent.com/73056004/116651518-f3d61380-a9b5-11eb-8625-32b3780ab342.png)
![image](https://user-images.githubusercontent.com/73056004/116651533-f9cbf480-a9b5-11eb-8a89-353c938d9b37.png)
Exact solution
![image](https://user-images.githubusercontent.com/73056004/116651553-081a1080-a9b6-11eb-9343-4bbe9be9529d.png)
![image](https://user-images.githubusercontent.com/73056004/116651559-0cdec480-a9b6-11eb-8e74-81690a98567e.png)
It is strange that both of them get a problem at t=0.5.
If you know the reason, please tell me.
Thanks a lot.

**Comment by lululxvi** [comment]
It is hard to read the figure in 3D. Do you try to simulate for a small time domain to check whether it works?

**Comment by xy1015874723** [comment]
@lululxvi Not yet, I will try it , thank you.
However, I met a problem recently, could you please give me some advice?
I want to use DeepXDE to solve  stress  and diffusion coupling Partial differential equations simultaneously.
Here are the information about the 2 equations.
![image](https://user-images.githubusercontent.com/73056004/117527570-95391700-afff-11eb-8a22-8a5cab9b2542.png)
And I have tried it, but I am not sure whether DeepXDE can solve this problem.
Here are my code and   I met an error.
![image](https://user-images.githubusercontent.com/73056004/117527834-6c198600-b001-11eb-8b3d-5e8f6f3e845e.png)
![image](https://user-images.githubusercontent.com/73056004/117527861-9ff4ab80-b001-11eb-961b-7f941dc92fc9.png)
I believed that DeepXED can solve my problem, and  the error happened just because something goes wrong with my code.
I'm dying for solve my problem by DeepXDE, please give me some advice.
Thanks a lot.

**Comment by xy1015874723** [comment]
@lululxvi Not yet. I will try it, thanks a lot

**Comment by xy1015874723** [comment]
@lululxvi  really sorry to bother you again.
I am using DeepXDE  to slove Partial differential equations simultaneously. 
Here  are my PDEs
![image](https://user-images.githubusercontent.com/73056004/118121380-fad43b80-b423-11eb-933e-7848b2c80483.png)
However, the results seems unsatisfying. I guess that maybe because I define bcs incorrectly.(maybe other reasons) Here is my code.
![image](https://user-images.githubusercontent.com/73056004/118122335-4cc99100-b425-11eb-9531-a647428a439c.png)
![image](https://user-images.githubusercontent.com/73056004/118122362-53f09f00-b425-11eb-879b-10c4a7d1ac58.png)
Could you please give me some pivotal advice? 
Thanks a lot.

**Comment by lululxvi** [comment]
The scaling is not good. Try to rescale your problem, such that x is from [0, 1], and other parameters are of order 1. See FAQ "Q: I failed to train the network or get the right solution, e.g., the training loss is large."

**Comment by xy1015874723** [comment]
@lululxvi Thanks a lot.
I am trying to rescale my problem. I think this bc can be used as  Neumann bc 
![image](https://user-images.githubusercontent.com/73056004/119661849-4c88b700-be63-11eb-886a-3553eea3311d.png)
where nu=0.3 I/F = J_s
and my code
![image](https://user-images.githubusercontent.com/73056004/119662429-f6684380-be63-11eb-878b-46743e1ddc20.png)
![image](https://user-images.githubusercontent.com/73056004/119662526-126be500-be64-11eb-8da2-210fa2d46f88.png)
however I got this error
![image](https://user-images.githubusercontent.com/73056004/119662787-5959da80-be64-11eb-992b-e05d99eaad47.png)
I thought  this because I mistake defining bc.  Do you know the reason or I have to use operatorBC ?
Please give me some advice.
Thanks a lot

**Comment by xy1015874723** [comment]
@lululxvi Thank you very much. With your advice, I rescale my PDE and set x from [0, 1] by dimensionless method, and finally get right results. However, I'm a little confused that when I use DeepXDE if the spatiotemporal domain is very small/large like my PDE above, do I have to rescale my model to common dimension in advance and then give it to DeepXDE ?
Look forward to your reply. Thanks again.

**Comment by lululxvi** [comment]
Yes, you need to rescale, which is required in deep learning training. You can either rescale by yourself, or directly in DeepXDE, see an example at https://github.com/alirezayazdani1/SBINNs

**Comment by xy1015874723** [comment]
@lululxvi Thanks. I have read it, and  it is really a good job. And I found a interesting phenomenon.
![image](https://user-images.githubusercontent.com/73056004/123505881-073eeb80-d694-11eb-9ecc-4e5d0130cb2f.png)
From the training result, the training points seems  to be sparse in some domain(especially  at lower right corner)
I think you must be know the reason, and how can l make the training points become more uniform.
Thanks again.

**Comment by lululxvi** [comment]
Is it ture that it is sparse in the corner? It is hard to see in 3D. You may plot the locations of points in 2D from train.dat for a double check.

**Comment by xy1015874723** [comment]
Oh, sorry. The 3D plot is puzzling. You are right. 
By the way,  the loss of one bc does not change in the training process, like this
![image](https://user-images.githubusercontent.com/73056004/123634002-79f2c700-d84c-11eb-9969-4a1f615c4188.png)
in column 6, and the results are wrong. 
Are  inappropriate rescale one of the reasons for this error?
Thanks a lot

**Comment by lululxvi** [comment]
Yes, the column 6 is always 1e-6, which is very wired. There could be something wrong with the code.

**Comment by xy1015874723** [comment]
Hi, Mr Lu
I found a phenomenon that when I handled the PDE in the form
![image](https://user-images.githubusercontent.com/73056004/132122556-b7b2938b-f693-4962-8fba-ffd957973059.png)
the loss would be very large, and got wrong results.
while I re-write the PDE like this 
![image](https://user-images.githubusercontent.com/73056004/132122638-1226f992-03f4-4736-addc-85c45d674927.png)
DeepXDE can found satisfying results.
Here x from 0 to 1
Could you please explain it for me?
Thanks a lot.

**Comment by lululxvi** [comment]
When x=0 or near 0, 2/x blows up.

**Comment by xy1015874723** [comment]
Thanks, I think so as well, and sometimes when I change some parameters (e.g., learning rate and loss weights), the predicted results always keep unchanged. Is that normal?
Thanks again.

**Comment by xy1015874723** [comment]
Thanks. I think so as well, and in this model
![L~IEJXEM8QDM9UG0WRC6F$Q](https://user-images.githubusercontent.com/73056004/132937918-3a75dde2-221a-4299-81ad-420063d5df2e.png)
the loss of ic in column is much larger than others. However, the results are acceptable but not perfect. Do you have some advice 
to improve the accuracy and make the loss of ic become normal?
Thanks again.

**Comment by lululxvi** [comment]
What is the last loss?

**Comment by xy1015874723** [comment]
the loss of ic

**Comment by lululxvi** [comment]
To make the NN match IC better, you can use a larger weight for the IC. You can also enforce IC exactly as a hard constraint. See FAQ.
